### PR TITLE
Add channels index

### PR DESCRIPTION
**Pull request by KyleMaas**
Tried adding a channel index.  This actually seems to work pretty well.

### CODE REVIEW

**Comment by arj03** [comment]
Yes, nice. Need to test what this does for onboarding performance.

**Comment by KyleMaas** [comment]
Without channel index:

```
full feeds: 513.962158203125 ms
partial feeds: 4963.927001953125 ms
```

With index:

```
full feeds: 548.82275390625 ms
partial feeds: 4259.909912109375 ms
```

**Comment by KyleMaas** [comment]
Pulling channel list on Channels using Overall Popular (which queries all posts with a channel) vs. All Alphabetically (which uses the index):

```
channel list: 35.903076171875 ms
channel list: 0.638916015625 ms
```

This test was done with following both of your accounts so it would have more channel data to work with.

**Comment by KyleMaas** [comment]
See https://github.com/arj03/ssb-browser-demo/pull/197 for uses

**Comment by arj03** [comment]
Yeah I think we can do this. We need a way to disable the index because not all browser apps want to use this index. I tested indexing and its 850 ms outside the browser so that seems ok. I have a few comments on the index that would be good to get in before we merge this.

File: indexes/channels.js [hunk 1]
```diff
@@ -0,0 +1,109 @@
+const bipf = require('bipf')
+const pull = require('pull-stream')
+const pl = require('pull-level')
+const jsonCodec = require('flumecodec/json')
+const Plugin = require('ssb-db2/indexes/plugin')
+
+const isFeed = require('ssb-ref').isFeed
```
Comment (arj03) [inline indexes/channels.js:7]: not needed

File: indexes/channels.js [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+// 1 index:
```
Comment (arj03) [inline indexes/channels.js:9]: just remove these 2 comment lines

File: indexes/channels.js [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+  const bChannel = Buffer.from('channel')
+  const bPost = Buffer.from('post')
+
+  let batch = []
+
+  const name = 'channels'
+  const { level, offset, stateLoaded, onData, writeBatch } = Plugin(
+    dir,
+    name,
+    3,
```
Comment (arj03) [inline indexes/channels.js:25]: start with 1

File: indexes/channels.js [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+        key: channel,
+        value: channels[channel]
+      })
+    }
+
+    return batch.length
+  }
+  
+  function updateChannelData(channel) {
+    channels[channel] = { id: channel }
```
Comment (arj03) [inline indexes/channels.js:69]: if you have a dict here, why not add a count? Might be useful to find most used channels?

File: indexes/channels.js (hunk 1 above)
Comment (KyleMaas) [reply indexes/channels.js:7]: isFeed could do, but jsonCodec and Plugin are used further down.

Comment (arj03) [reply indexes/channels.js:7]: Right, the comment was only for isFeed

**Comment by KyleMaas** [comment]
Have not tested that version yet, but that should include all the changes.

**Comment by KyleMaas** [comment]
Tried it, and it works pretty well.  Want to make one more change so that `##channel` is counted as just `channel`.  If we're going to keep counts, it makes a difference.

**Comment by arj03** [comment]
Yes, no rush with this. Lets get this index right before merging :)

File: indexes/channels.js [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+      })
+    }
+
+    return batch.length
+  }
+  
+  function updateChannelData(channel) {
+    if (!channels[channel])
+      channels[channel] = { id: channel, count: 0 }
+    if (!channels[channel].count) // For compatibility with existing index, if it exists.
```
Comment (arj03) [inline indexes/channels.js:66]: I guess this is from your existing index? It's better if you just nuke it so we don't need that code here.

**Comment by KyleMaas** [comment]
Fixed stray leading '#'s, removed compatibility code.  Checked it, and "patchbook" now shows a count of 3 where before you had to add "patchbook" and "#patchbook" together to get that.

**Comment by KyleMaas** [comment]
I'm pretty happy with the way this works at this point.  Stripping out leading '#'s means I can remove some of the duplicate code from the channel-index branch of ssb-browser-demo.

**Comment by arj03** [comment]
Tested this. Did a full sync of my browser profile and the sync time did not noticeable change. Are the changes to ssb-browser-demo in the PR ready? It's no rush, just don't want to merge this before :)

**Comment by KyleMaas** [comment]
Yes.  I left it in Draft status until this was ready, but it worked quite well when I last tested it.

**Comment by arj03** [comment]
Sweet. Will release this as 8.1.0 in a minute or two :)

**Comment by KyleMaas** [comment]
Awesome.  You need me to update package.json and npm-shrinkwrap.json on the ssb-browser-demo branch to fix the merge conflicts, or do you want to do that?

**Comment by arj03** [comment]
Can you do that @KyleMaas ?

**Pull request merged**